Fix ESPHome color temperature precision for light entities

danielkent-net opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
<!-- ## Breaking change -->
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change
This PR changes the behavior of ESPHome `light` entities when setting color temperature. Currently, the color temperature can only be set in whole-number mireds, which cannot map to the full range of Kelvin color temperature values.

Since ESPHome entities both transmit and store color temperature in floating point mireds, we can use this property to convert this value to and from Kelvin more accurately, and avoid many of the issues that result from rounding the value.

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #91392 fixes #89846
- This PR is related to [esphome issue 2186](https://github.com/esphome/feature-requests/issues/2186)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

danielkent-net committed:
Accurately get/set Kelvin color temp for esphome entities

In components/esphome/light.py:
 - Add ADDR_COLOR_TEMP_KELVIN to list of parsed attributes
 - Import mired to kelvin conversion util code
 - If ATTR_COLOR_TEMP_KELVIN is set in service call, calculate target mired
   value in floating point. Fall back to ATTR_COLOR_TEMP if _KELVIN not available
 - Add {min,max}_color_temp_kelvin properties, computed from floating point mired,

danielkent-net commented:
Based on the current changes, I can verify that floating point mireds are now being sent to ESPHome light devices when changing them via the color temperature slider in the Web UI:
![image](https://user-images.githubusercontent.com/129895318/232078562-38569fc8-045e-4df0-b2e9-a45eb451f1d8.png)

However, I am still seeing some issues, which is why I am keeping this PR as a draft for now:

 - The Kelvin value shown in the Web UI usually changes slightly between when it's first clicked and after it's sent to the ESPHome device.
 - The color temperature still cannot be set to the absolute upper bound using the slider; in my current test setup with a configured 3000K to 6000K range, I can only set the maximum to 5999K. It's better than it was before, but still annoying.

danielkent-net commented:
As a related sidenote, the [documentation for the Home Assistant `light` entity](https://www.home-assistant.io/integrations/light/) seems a bit outdated; it still mentions `color_temp` and `kelvin` properties in the service call, while [in the code these properties are marked as deprecated](https://github.com/home-assistant/core/blob/dev/homeassistant/components/light/__init__.py#L204) and to favor the `color_temp_kelvin` property instead. I can open a new issue on the home-assistant.io repo for that one.

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net committed:
Define color_temp_kelvin for ESPHome lights

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net commented:
Finally figured out the remaining issue; I needed to override the `color_temp_kelvin` property, and the parent function looks for a property that doesn't exist, so it used the built-in mired to Kelvin conversion that loses precision. With the latest change, this precision is maintained all the way to the UI.

There is one remaining issue, but it's out of scope of this PR: the function `color_temperature_mired_to_kelvin` in [`util/color.py:619`](https://github.com/home-assistant/core/blob/dev/homeassistant/util/color.py#L619) uses a floor function. This causes any values with fractional mired values to be off by one if the fractional part is greater than 0.5. I think it would be more appropriate to round, but changing that might break other integrations.

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

bdraco commented:
> There is one remaining issue, but it's out of scope of this PR: the function `color_temperature_mired_to_kelvin` in [`util/color.py:619`](https://github.com/home-assistant/core/blob/dev/homeassistant/util/color.py?rgh-link-date=2023-04-18T17%3A33%3A33Z#L619) uses a floor function. This causes any values with fractional mired values to be off by one if the fractional part is greater than 0.5. I think it would be more appropriate to round, but changing that might break other integrations.

You could open another PR for this

bdraco committed:
Merge branch 'dev' into esphome_color_temp_precision

bdraco commented:
Testing this

before

<img width="520" alt="Screenshot 2023-05-05 at 5 02 21 PM" src="https://user-images.githubusercontent.com/663432/236576025-968a3ebe-c658-4ce4-8df3-18eda7bf30ac.png">
<img width="486" alt="Screenshot 2023-05-05 at 5 02 18 PM" src="https://user-images.githubusercontent.com/663432/236576027-b3381473-7def-4e96-8548-8c2385a20e1c.png">

bdraco commented:
After
<img width="560" alt="Screenshot 2023-05-05 at 5 03 25 PM" src="https://user-images.githubusercontent.com/663432/236576097-c8e1f413-543a-4cc4-a721-a9e8cc4fbcba.png">
<img width="504" alt="Screenshot 2023-05-05 at 5 03 21 PM" src="https://user-images.githubusercontent.com/663432/236576100-650d85b7-01a6-4d99-a9d1-facd67067e1c.png">

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -342,6 +352,11 @@ def color_temp(self) -> int:
+    @property
+    def color_temp_kelvin(self) -> int:
+        """Return the CT color value in Kelvin."""
+        return mired_to_kelvin(self._state.color_temperature)
```
```suggestion
        return round(1000000 / self._state.color_temperature)
```

danielkent-net replied:
Internal function _mired_to_kelvin using the same computation added in commit 879575d1385c06588fbe2d86f39f22a71506e791

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -391,3 +406,13 @@ def min_mireds(self) -> int:
+    @property
+    def min_color_temp_kelvin(self) -> int:
+        """Return the warmest color_temp that this light supports, in Kelvin."""
+        return mired_to_kelvin(self._static_info.max_mireds)
```
```suggestion
       return round(1000000 / self._static_info.max_mireds)
```

danielkent-net replied:
Internal function _mired_to_kelvin using the same computation added in commit 879575d1385c06588fbe2d86f39f22a71506e791

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -391,3 +406,13 @@ def min_mireds(self) -> int:
+    @property
+    def max_color_temp_kelvin(self) -> int:
+        """Return the coldest color_temp that this light supports, in Kelvin."""
+        return mired_to_kelvin(self._static_info.min_mireds)
```
```suggestion
        return round(1000000 / self._static_info.min_mireds)
```

danielkent-net replied:
Internal function _mired_to_kelvin using the same computation added in commit 879575d1385c06588fbe2d86f39f22a71506e791

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -24,6 +25,9 @@
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.color import (
+    color_temperature_mired_to_kelvin as mired_to_kelvin,
+)
```
```suggestion
```

danielkent-net replied:
Import removed in commit 879575d1385c06588fbe2d86f39f22a71506e791

bdraco commented:
With suggested changes above

<img width="528" alt="Screenshot 2023-05-05 at 5 06 05 PM" src="https://user-images.githubusercontent.com/663432/236576525-80b863c3-8ad1-474e-8420-195ca2cb3961.png">
<img width="534" alt="Screenshot 2023-05-05 at 5 06 01 PM" src="https://user-images.githubusercontent.com/663432/236576526-79024eb0-8524-4af3-a0d5-5cd4081e0019.png">

bdraco commented:
I'd suggestion making a esphome specific `mired_to_kelvin` function instead. We can merge that here and than you can pursue a change to the helper if you want. If the helper change gets merged, you can come back and take it the esphome helper and use the core one later.

danielkent-net committed:
Moved mired_to_kelvin to internal helper function

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net commented:
> I'd suggestion making a esphome specific `mired_to_kelvin` function instead. We can merge that here and than you can pursue a change to the helper if you want. If the helper change gets merged, you can come back and take it the esphome helper and use the core one later.

I appreciate the feedback! I decided to take this approach; there is now a `_mired_to_kelvin` function defined within `light.py` (following the convention of other internal functions such as `_color_mode_to_ha` and `_filter_color_modes`) that is then used for all the instances where mired_to_kelvin was previously used.

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -342,6 +358,11 @@ def color_temp(self) -> int:
         return round(self._state.color_temperature)
 
+    @property
+    def color_temp_kelvin(self) -> int:
```
```suggestion
    @esphome_state_property
    def color_temp_kelvin(self) -> int:
```

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -391,3 +412,13 @@ def min_mireds(self) -> int:
     def max_mireds(self) -> int:
```
Drop the `min_mireds` and `max_mireds` as well since they are deprecated
https://github.com/home-assistant/core/blob/053eaad2bd09ac1b656be5ae22b379f36891a295/homeassistant/components/light/__init__.py#L206

danielkent-net replied:
I disagree with this suggestion; I would argue that removing a deprecated property should be its own PR. The presence of a deprecated property is not related to the bug in question, and changing this might break some setups. Also, there are several other light integrations that still use the min/max mired property, so I see no reason to remove the property in a bugfix PR.

bdraco replied:
I don't agree so I marked this as `second-opinion-wanted` to give another member a chance to weigh in.

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -212,8 +222,14 @@ async def async_turn_on(self, **kwargs: Any) -> None:
         if (transition := kwargs.get(ATTR_TRANSITION)) is not None:
             data["transition_length"] = transition
 
```
You'll need to fix the `min_ct` and `max_ct` above as well

```
-                    min_ct = self.min_mireds
-                    max_ct = self.max_mireds
+                    min_ct = self.min_color_temp_kelvin
+                    max_ct = self.max_color_temp_kelvin
```

bdraco replied:
This would need to be converted

bdraco replied:
Might as well leave it since its in the static info anyways and thats not going to change.

emontnemery commented:
I'm setting this PR to draft because there are unresolved review comments. Please mark it ready for review when it's been updated 👍

danielkent-net committed:
Add @esphome_state_property decorator to color_temp_kelvin

Add @esphome_state_property decorator to color_temp_kelvin function in homeassistant/components/esphome/light.py

Co-authored-by: J. Nick Koston <nick@koston.org>

danielkent-net committed:
Merge branch 'home-assistant:dev' into esphome_color_temp_precision

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net commented:
Setting this back to ready to review since we need a resolution on the point of contention. I still disagree with the last proposed change for a couple reasons:

1. Removal of deprecated functions and properties is a breaking change, which I want to avoid for this bugfix PR
2. The code changes required to use kelvin under the hood for certain features are not trivial, and could break setups if not implemented and tested carefully.

bdraco commented:
> * Removal of deprecated functions and properties is a breaking change, which I want to avoid for this bugfix PR

Its not a breaking change because the underlying light entity will still provide the value for backwards compatiblity

https://github.com/home-assistant/core/blob/dev/homeassistant/components/light/__init__.py#L930

danielkent-net committed:
Resolve merge conflict with new static info update

danielkent-net committed:
Change white+color light calc property lookup for min/max ct

danielkent-net commented:
@bdraco I see you opened another PR #94973 to handle this issue. I'm still willing to finish this one out, including integration with the changes that were recently made that can obviate some of the added property functions. I think what I've done should work; I can test on actual hardware later today.

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -220,8 +231,14 @@ async def async_turn_on(self, **kwargs: Any) -> None:
+        if (color_temp_k := kwargs.get(ATTR_COLOR_TEMP_KELVIN)) is not None:
+            # Do not use kelvin_to_mired here to prevent precision loss
+            data["color_temperature"] = 1000000.0 / color_temp_k
+        elif (color_temp := kwargs.get(ATTR_COLOR_TEMP)) is not None:
```
Untested (traveling), but I think you can drop everything but ATTR_COLOR_TEMP_KELVIN since the underlying light platform will already do the conversion to kelvin anytime the legacy ATTR_COLOR_TEMP is passed

bdraco replied:
See preprocess_turn_on_alternatives in light/__init__.py

bdraco replied:
https://github.com/home-assistant/core/blob/31f845bfe03aa6e8284deeb6d8f81f2effe473a7/homeassistant/components/light/__init__.py#L321

danielkent-net replied:
Looking through parts of the template light class, it appears as though using kelvin will override any mireds properties we set manually. So, in theory, removing the ATTR_COLOR_TEMP shouldn't do much.

I did think of and test one thing: if we call the service `light.turn_on` and use mireds temperature, it will convert to Kelvin and back, which does lead to small rounding errors (less than a mired from my limited testing). In practice this _shouldn't_ make too much of a different, and I have to imagine the number of users who would be significantly impacted is vanishingly small, but it's something we should keep in mind. I'll go ahead and make the change, but we should make sure the impact is as minor as I think it is.

bdraco replied:
Let's take it out.  We can fix the rounding on the light platform in a followup

bdraco replied:
I double checked the frontend as well. Its using kelvin all the way though so only anyone still using mireds would be affected by the double conversion which is an existing issue so its seems ok

https://github.com/home-assistant/frontend/blob/eb552530e2514a51c995a4e750e8ceff8684342b/src/dialogs/more-info/components/lights/light-color-temp-picker.ts#L112

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -388,3 +411,10 @@ def _on_static_info_update(self, static_info: EntityInfo) -> None:
+        self._attr_min_color_temp_kelvin = _mired_to_kelvin(static_info.max_mireds)
+        self._attr_max_color_temp_kelvin = _mired_to_kelvin(static_info.min_mireds)
+
+    @property
```
Is this leftovers from the merge conflict?

danielkent-net replied:
No, this is intentional. By adding the `_attr_{min,max}_color_temp_kelvin` class properties here, we don't need to implement the base class `{min,max}_color_temp_kelvin()` functions, which simplifies the proposed code quite a bit. I tested with and without these properties; with the properties it works as expected. Without, the upper/lower bounds of the color temperature selector become inaccurate.

bdraco replied:
I was referring to the `effect_list` property 
```
    def effect_list(self) -> list[str]:
        """Return the list of supported effects."""
        return self._static_info.effects
```

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -388,3 +411,10 @@ def _on_static_info_update(self, static_info: EntityInfo) -> None:
         self._attr_max_mireds = round(static_info.max_mireds)
+        self._attr_min_color_temp_kelvin = _mired_to_kelvin(static_info.max_mireds)
+        self._attr_max_color_temp_kelvin = _mired_to_kelvin(static_info.min_mireds)
+
+    @property
+    def effect_list(self) -> list[str]:
+        """Return the list of supported effects."""
+        return self._static_info.effects
```
```suggestion
```

danielkent-net replied:
I see what you mean now. These lines should be safe to remove.

danielkent-net committed:
Apply suggestions from code review

Remove effect_list from esphome light entity in favor of template class implementation

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -220,8 +231,14 @@ async def async_turn_on(self, **kwargs: Any) -> None:
+            data["color_temperature"] = 1000000.0 / color_temp_k
+        elif (color_temp := kwargs.get(ATTR_COLOR_TEMP)) is not None:
             data["color_temperature"] = color_temp
+        if (color_temp := kwargs.get(ATTR_COLOR_TEMP)) is not None or (
+            color_temp := kwargs.get(ATTR_COLOR_TEMP_KELVIN)
+        ) is not None:
```
```suggestion
```

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -220,8 +231,14 @@ async def async_turn_on(self, **kwargs: Any) -> None:
+        if (color_temp_k := kwargs.get(ATTR_COLOR_TEMP_KELVIN)) is not None:
+            # Do not use kelvin_to_mired here to prevent precision loss
+            data["color_temperature"] = 1000000.0 / color_temp_k
+        elif (color_temp := kwargs.get(ATTR_COLOR_TEMP)) is not None:
```
```suggestion
```

danielkent-net committed:
Remove ATTR_COLOR_TEMP option in favor of base class methods for conversion

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

bdraco commented:
I'll need to retest with a esphome light when I get home later today.

Might be friday ... its 2:30 in the morning here and I'm still up and have to catch an early flight back to the States.

bdraco commented:
Thanks @danielkent-net

bdraco commented:
Just put this on a test system an noticed it needs a guard for 0
```
2023-06-21 14:42:42.200 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Mid Garage for esphome
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 390, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/esphome/__init__.py", line 453, in async_setup_entry
    await entry_data.async_update_static_infos(hass, entry, infos)
  File "/usr/src/homeassistant/homeassistant/components/esphome/entry_data.py", line 256, in async_update_static_infos
    callback_(entity_infos)
  File "/usr/src/homeassistant/homeassistant/components/esphome/__init__.py", line 711, in async_list_entities
    entity = entity_type(entry_data, component_key, info, state_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/esphome/__init__.py", line 799, in __init__
    self._on_static_info_update(entity_info)
  File "/usr/src/homeassistant/homeassistant/components/esphome/light.py", line 400, in _on_static_info_update
    self._attr_min_color_temp_kelvin = _mired_to_kelvin(static_info.max_mireds)
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/esphome/light.py", line 107, in _mired_to_kelvin
    return round(1000000 / mired_temperature)
                 ~~~~~~~~^~~~~~~~~~~~~~~~~~~
ZeroDivisionError: float division by zero
```

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -388,3 +405,5 @@ def _on_static_info_update(self, static_info: EntityInfo) -> None:
         self._attr_effect_list = static_info.effects
         self._attr_min_mireds = round(static_info.min_mireds)
         self._attr_max_mireds = round(static_info.max_mireds)
+        self._attr_min_color_temp_kelvin = _mired_to_kelvin(static_info.max_mireds)
+        self._attr_max_color_temp_kelvin = _mired_to_kelvin(static_info.min_mireds)
```
```suggestion
        if ColorMode.COLOR_TEMP in supported:
            self._attr_min_color_temp_kelvin = _mired_to_kelvin(static_info.max_mireds)
            self._attr_max_color_temp_kelvin = _mired_to_kelvin(static_info.min_mireds)
```

danielkent-net committed:
Update homeassistant/components/esphome/light.py

Check for color temp support before setting Kelvin color attributes

Co-authored-by: J. Nick Koston <nick@koston.org>

danielkent-net commented:
I noticed the main implementation of mired_to_kelvin didn't have a zero division check so I didn't add one to `_mired_to_kelvin()` initially. Should we choose a sane default instead? The default range in HA light entities is 2000-6500K, so we could do something like:

`return round(1000000/mired_temperature) if mired_temperature > 0 else 6500`

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

bdraco commented:
> I noticed the main implementation of mired_to_kelvin didn't have a zero division check so I didn't add one to `_mired_to_kelvin()` initially. Should we choose a sane default instead? The default range in HA light entities is 2000-6500K, so we could do something like:
> 
> 
> 
> `return round(1000000/mired_temperature) if mired_temperature > 0 else 6500`

Since we don't know if it's the minimum or maximum value its effectively GIGO at this point and it's probably best to return the original value (0) so we don't have to figure out later where the data was mutated on us if there is a bug.

bdraco commented:
https://github.com/home-assistant/core/pull/95029 will add some basic tests for this platform

Won't be able to test today as I won't get home till tomorrow now as my original flight canceled and I have a few stops now

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net committed:
Add sanity check for _mired_to_kelvin for esphome to avoid divide by zero

bdraco commented on homeassistant/components/esphome/light.py:
```diff
@@ -98,6 +98,22 @@ async def async_setup_entry(
+    the original value to avoid a divide by zero.
+    """
+    return (
+        round(1000000 / mired_temperature)
```
For the HA codebase, please unpack ternarys into an if block when they multiple lines

bdraco commented:
Managed to pickup a different flight so might be able to test later but at worst I'll be able to test tomorrow.  Still plenty of time to get this merged before beta on Wednesday

danielkent-net committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net committed:
Refactor _mired_to_kelvin to match HA convention

bdraco committed:
Merge branch 'dev' into esphome_color_temp_precision

bdraco commented:
One of the new tests will likely need to be adjusted for the behavior change

sorry about the conflicts

bdraco commented:
One of the new tests will likely need to be adjusted for the behavior change

sorry about the conflicts

bdraco committed:
Merge branch 'dev' into esphome_color_temp_precision

danielkent-net commented:
Unfortunately it looks like the functionality that automatically derives the min/max mireds from Kelvin around `components/light/__init__.py:928` still uses the floor function, and that's well out of the esphome light scope.

The only way I can think to get around this is to override the `capability_attributes` function, first calling the superclass function to get the base dictionary, then if Kelvin properties exist repopulate the `ATTR_{MIN,MAX}_MIREDS` properties with our "correct" implementation. Something like:

```
   @property
    @esphome_state_property
    def capability_attributes(self) -> dict[str, Any]:
        """Return capability attributes.
           This function takes the base class attributes and corrects any
           incorrect calculation of the color temperature bounds in mireds."""
        data = super(LightEntity, this).capability_attributes()
        if ATTR_MIN_COLOR_TEMP_KELVIN in data:
            data[ATTR_MAX_MIREDS] = _kelvin_to_mireds(data[ATTR_MIN_COLOR_TEMP_KELVIN])
        if ATTR_MAX_COLOR_TEMP_KELVIN in data:
            data[ATTR_MIN_MIREDS] = _kelvin_to_mireds(data[ATTR_MAX_COLOR_TEMP_KELVIN])
        return data
```

IMO it's an inelegant solution, but until the base class implementation is fixed, we're going to still see off-by-one errors.

bdraco commented:
Let's adjust the test here and get this merged (assuming testing is ok)

Than the light class can be adjusted and we can rework this again

bdraco committed:
Merge branch 'dev' into esphome_color_temp_precision

bdraco committed:
fix tests

bdraco commented:
I just realized the test is wrong since esphome sends the full float and the test doesn't.

I'll fix the test and merge this

bdraco commented:
Thanks @danielkent-net

bdraco committed:
Merge branch 'dev' into esphome_color_temp_precision
